[GeoMechanicsApplication] Add settlement test case 3

avdg81 opened the pull request:
**📝 Description**
Added a settlement test case that corresponds to Test Case 3 provided by the Embankment Suite team. This test case checks a soil column, where the phreatic line is located 10 m below the soil surface.

Also added functionality to easily generate settlement history plots and stresses-over-the-depth plots. This makes it much easier to compare results obtained with Kratos with other known results.

**🆕 Changelog**
- The `FixWaterPressuresAbovePhreaticLineProcess` now uses the nodal coordinates (which can be either the initial positions [when `"move_mesh"` is set to `false`] or the updated positions [when `"move_mesh"` is set to `true`]). Consequently, some test results had to be updated.
- Added a new test case to `tests/test_dsettlement_validation.py` that corresponds to Test Case 3 of the Embankment Suite team. It basically consists of a soil column, where the phreatic line is located 10 m below the soil surface. This new test case can also generate the settlement history plot as well as several stresses-over-the-depth plots. This behavior is controlled by an environment variable, which has been documented as well.
- For some other test cases, the requested output has been extended such that we can generate plots for them.
- Added many text files with reference results that have been produced with D-Settlement 23.2.1. The reference data is being used when generating plots.
- Added utility functions for making settlement history plots and stress plots.
- Added utility functions for converting a value from one unit to another.
- Added several test helper functions, including one to read data points from a text file, and one to extract settlement history data. Also modified the function that extracts nodal result data for the given time. It will now fail when the requested output data is not present for the given node(s).
- When running a settlement analysis (using `python_scripts/run_geo_settlement.py`) the current working directory is changed to the given working directory prior to calling `RunStage`. Afterwards, it's changed back to the original working directory. In this way, any client code no longer needs to change it explicitly.
- Fixed a typo.

avdg81 committed:
Added input files of test case 3

Slightly modified versions of David's original input files. Further
cleanup is still needed.

I've also started to document the test case.

avdg81 committed:
Reformatted the material parameters files

Also corrected the model part's name.

avdg81 committed:
Reformatted the project parameters files

avdg81 committed:
Added code to run the new test case

Also corrected the path to the abc model library.

avdg81 committed:
Cleaned up the material properties

- Removed redundant properties.
- Use a more consistent notation for floating-point numbers.

avdg81 committed:
Removed a duplicated model part

'K0_model_part' was a duplicate of 'Soil'. The former has been removed, and its usages now refer to the latter.

Furthermore, the nodes listed by 'Soil' have been sorted, and the empty section 'SubModelPartTables' has been removed.

avdg81 committed:
Removed another duplicated model part

'gravity' was a duplicate of 'Soil'. The former has been removed, and its usages now refer to the latter.

avdg81 committed:
Removed yet another duplicated model part

'Head_line0' was a duplicate of 'Soil'. The former has been removed, and its usages now refer to the latter.

avdg81 committed:
Removed an unused model part

Model part 'Head_line1' was defined, but never used.

avdg81 committed:
Removed several empty sub model part sections

avdg81 committed:
Renamed a model part

'TopWaterPressure0' => 'TopWaterPressure'

avdg81 committed:
Renamed a model part

'PhreaticLineContactStress0' => 'PhreaticLineContactStress'

avdg81 committed:
Corrected some material properties

The permeability values as well as the dynamic viscosity were apparently defined using days as the time unit, whereas seconds should have been used. That has been corrected.

avdg81 committed:
Cleaned up stage 1 further

- Make sure that the initial field for the water pressure is defined prior to defining the water pressure boundary conditions at the top and bottom edges. Since the initial field includes the top and bottom edges, the order is important here. The more specific conditions should come last.
- Removed the application of the normal contact stress since it had no effect. There is no water above the soil surface.
- Removed a few redundant "value" specifications.
- Reordered the processes that set the boundary conditions: first define the ones applicable to the displacement field, then the ones applicable to the water pressure.

avdg81 committed:
Cleaned up stage 2 further

- Removed the application of the normal contact stress since it had no effect. There is no water above the soil surface. Also removed the corresponding model part.
- Removed a few redundant "value" specifications.
- Reordered the processes that set the boundary conditions: first define the ones applicable to the displacement field, then the ones applicable to the water pressure field.
- Removed the process that extrapolates the total stresses and Cauchy stresses from the integration points to the nodes.

avdg81 committed:
Cleaned up stage 3 further

- Removed the application of the normal contact stress since it had no effect. There is no water above the soil surface. Also removed the corresponding model part.
- Removed a few redundant "value" specifications.
- Reordered the processes that set the boundary conditions: first define the ones applicable to the displacement field, then the ones applicable to the water pressure field.
- Removed the process that extrapolates the total stresses and Cauchy stresses from the integration points to the nodes.

avdg81 committed:
Cleaned up stage 4 further

- Removed the application of the normal contact stress since it had no effect. There is no water above the soil surface. Also removed the corresponding model part.
- Removed a few redundant "value" specifications.
- Reordered the processes that set the boundary conditions: first define the ones applicable to the displacement field, then the ones applicable to the water pressure field.
- Removed the process that extrapolates the total stresses and Cauchy stresses from the integration points to the nodes.

avdg81 committed:
Cleaned up stage 5 further

- Removed the application of the normal contact stress since it had no effect. There is no water above the soil surface. Also removed the corresponding model part.
- Removed a few redundant "value" specifications.
- Reordered the processes that set the boundary conditions: first define the ones applicable to the displacement field, then the ones applicable to the water pressure field.
- Removed the process that extrapolates the total stresses and Cauchy stresses from the integration points to the nodes.

avdg81 committed:
Make some plots for test case 3

To investigate why the test results differ from the results obtained with the "old" D-Settlement package, we decided to make a few plots to see what is going on.

avdg81 committed:
Improved the settlement plot

- Inverted the vertical axis to get a more natural view of the settlement history.
- Labeled the vertical axis 'Settlement'.
- Renamed a function.

avdg81 committed:
Moved reference settlement data to a file

avdg81 committed:
Moved more reference data to files

The water pressure values as well the effective vertical stress values after 10,000 days have been moved to files to make the test script cleaner. Also renamed the data series as well as the axis labels.

avdg81 committed:
Extended the stress plot after 10,000 days

The plot now also shows the results obtained with Kratos.

avdg81 committed:
Use `pathlib` to simplify path manipulation

avdg81 committed:
Split 'Sides' into 'LeftSide' and 'RightSide'

avdg81 committed:
The settlement plot is now created conditionally

When the environment variable `KRATOS_GEO_MAKE_TEST_PLOTS` is set to `on`, the settlement plot is produced. Otherwise, it isn't.

avdg81 committed:
Refactored how the settlement plot is made

avdg81 committed:
The settlement plot now also uses output of stage 4

avdg81 committed:
Extracted a function for making a settlement plot

avdg81 committed:
Results from "old" D-Settlement now include the row number

This will make updating the reference data easier.

avdg81 committed:
Make plot of water pressures after 100.1 days

avdg81 committed:
Make settlement plots for the dry and fully wet test cases

avdg81 committed:
Renamed a function

avdg81 committed:
Added a function that converts a duration from days to seconds

Also renamed the function that converts a duration from seconds to days

avdg81 committed:
Added two helper functions for fetching vertical stresses and water pressures

avdg81 committed:
Moved shifting the y coordinate to a separate function

avdg81 committed:
Consider the undeformed mesh when fixing the water pressure

The process that fixes the water pressures above the phreatic line used to consider the _updated_ y coordinate (i.e. taking into account the settlement so far). Now it considers the nodal positions of the undeformed mesh.

Other changes include:
- The settlement plot of the Kratos results also uses the y coordinates of the undeformed mesh.
- Updated the reference data of test case 3.
- Extracted a function to get the vertical effective stresses.

avdg81 committed:
Completed the stress plot for time = 100.1 days

Added the reference data and extract the required results from the Kratos output files.

avdg81 committed:
Reuse the function that gets the nodal water pressures

avdg81 committed:
Added helper function for making stress plots

avdg81 committed:
Removed some commented out test code

avdg81 committed:
Moved the code that makes the stress plot at time = 100.1 days

avdg81 committed:
Updated the reference settlement data

These results were produced using the 1-D consolidation model from "old" D-Settlement. These results are nearly identical to the results produced with the 2-D consolidation model.

avdg81 committed:
Added a stress plot at time = 100 days

avdg81 committed:
Re-enabled the assertions

avdg81 committed:
Removed an unused condition

avdg81 committed:
Removed a duplicated table

Also changed the remaining table's identifier.

avdg81 committed:
Removed a duplicated model part for the line load

Also renamed the remaining model part.

avdg81 committed:
Merge remote-tracking branch 'origin/master' into geo/13745-settlement-test-case-3

avdg81 committed:
Introduced a Python script for making plots

So far, it only contains a function that extracts data points from a file.

avdg81 committed:
Moved two functions to the plot utilities

And moved another one back to the test script.

avdg81 committed:
Added a settlement plot for test case 4

avdg81 committed:
Only import the plot utilities when needed

avdg81 committed:
Extracted some common code

avdg81 committed:
Renamed a few function parameters

Also reformatted a Python script using `black`.

avdg81 committed:
Added a stress plot for test case 4 at time = 10,000 days

avdg81 committed:
Added stress plots for test case 4 at times 100 days and 100.1 days

avdg81 committed:
Improved the comments about the source of the reference data

avdg81 committed:
Fetching nodal results from GiD output no longer requires sorted node IDs

Internally, the function that fetches nodal results for the given time now uses a map from node ID to requested value. As such, it is no longer needed to pass sorted lists of node IDs.

Also added an assertion to test that the number of fetched nodal tensors matches the number of node IDs. Currently, this assertion fails for at least one test, which indeed appeared to be an error.

avdg81 committed:
Replaced both usages of `nodal_values_at_time2`

We now use the improved version of `nodal_values_at_time`. The temporary function `nodal_values_at_time2` has now been removed completely.

avdg81 committed:
Added water pressure results produced by D-Settlement

We were still missing the water pressure values over the depth for test case 3 at time = 100 days.

avdg81 committed:
Removed some duplicated test code

Added a list of expected settlement values at some given times. The actual check is now implemented only once.

avdg81 committed:
Use a "scoped" cwd when running a settlement calculation

avdg81 committed:
Moved a class and a function to the plot utilities

avdg81 committed:
Commented out a failing assertion

Aron and I agreed that he will look into this.

avdg81 committed:
Use the nodal position when fixing water pressures above the phreatic line

When undeformed meshes are considered (i.e. `MOVE_MESH_FLAG` == false), the nodal position equals the initial nodal position. When deformed meshes are considered (i.e. `MOVE_MESH_FLAG` == true), the nodal position accounts for any deformations that have occurred.

Also renamed two unit test cases to make clearer what they test.

avdg81 committed:
Corrected two expected values

For the case where we assume geometrically linear behavior, the water pressures above the phreatic line must be equal to zero, regardless of the computed displacement field.

avdg81 committed:
Removed the total displacement from the unit tests of `FixWaterPressuresAbovePhreaticLineProcess`

avdg81 committed:
Corrected the stress plot of test case 3 at time == 100 days

avdg81 committed:
Updated the reference water pressures of test case 3 at time == 100.1 days

avdg81 committed:
Added stress plots for test case 2

avdg81 committed:
Fixed a typo

avdg81 committed:
Corrected the reference results of test case 3 for time = 100.1 days

avdg81 committed:
Renamed two local variables

avdg81 committed:
Customize extraction of x and y from a line

avdg81 committed:
Renamed a few functions

avdg81 committed:
Moved unit conversion functions to a new Python script

avdg81 committed:
Inlined a variable

avdg81 committed:
Use data points rather than lists of x values and y values

avdg81 committed:
Added a function to convert from Pa to kPa

avdg81 committed:
Check that iterables have same lengths when `zip`ping them

avdg81 committed:
The y coordinates now need to be provided explicitly to the stress plot function

avdg81 committed:
Renamed a function

avdg81 committed:
Renamed a few more functions

avdg81 committed:
Store reference data in a dictionary

avdg81 committed:
Refactored three functions

avdg81 committed:
Use a lambda rather than a nested function

avdg81 committed:
Moved two functions to `test_helper.py`

avdg81 committed:
Reformatted a few Python scripts using `black`

avdg81 committed:
Added more checks for the settlement curve of test case 3

avdg81 committed:
Removed a redundant function call

WPK4FEM committed:
First attempt to documentation of case 3

avdg81 committed:
Documented how to let the test scripts generate plots

avdg81 committed:
Don't use double quotes inside of a double-quoted string

avdg81 committed:
Polished two `README.md` files

avdg81 committed:
Several minor improvements

WPK4FEM commented on applications/GeoMechanicsApplication/custom_processes/fix_water_pressures_above_phreatic_line.cpp:
```diff
@@ -42,9 +42,7 @@ FixWaterPressuresAbovePhreaticLineProcess::FixWaterPressuresAbovePhreaticLinePro
-        const auto current_coordinates =
-            rNode.GetInitialPosition() + rNode.FastGetSolutionStepValue(TOTAL_DISPLACEMENT);
-        if (current_coordinates[1] > mPhreaticLineTable(current_coordinates[0])) {
+        if (rNode.Coordinates()[1] > mPhreaticLineTable(rNode.Coordinates()[0])) {
```
This is quite a change:
The previous comparison always uses an updated position. Now it uses the original position in geometrical linear analysis and the updated position in geometrical non-linear ( so MOVE_MESH_FLAG: true ) analysis.
For me the new check is a more natural check. A computation is should either be geometrical linear or geometrical nonlinear in all aspects

Please list this change in the changelog of the pull request.

avdg81 replied:
I think it is listed now as the first bullet point of the change log. Please let me know if you feel the description is inadequate. Thanks.

WPK4FEM replied:
The description is adequate. Thank you.

WPK4FEM commented on applications/GeoMechanicsApplication/python_scripts/unit_conversions.py:
```diff
@@ -0,0 +1,10 @@
+    return duration_in_days * 24.0 * 60.0 * 60.0
+
+
+def Pa_to_kPa(stress_in_Pa):
```
I wonder whether a conversion from 1 unit to k or M (1000 or 1000000) with exactly the same functionality is more appropriate. E.g. for conversion from N to kN we need the same, but I would be tempted to write a new function as Pascals are in its name here.

avdg81 replied:
Good question. I'm not sure what would be the simplest solution here. One way could be to have simple wrapper functions that spell out the actual conversion being done (like `Pa_to_kPa`). The implementation could, however, be expressed in a more generic way.
I would suggest to revisit this question when the need arises. What do you think?

WPK4FEM commented on applications/GeoMechanicsApplication/tests/README.md:
```diff
@@ -0,0 +1,6 @@
+
+# Result plots
+
+Some tests can plot their results.  For instance, some settlement test cases (see `test_dsettlement_validation.py`) can plot the evolution of settlement over time.  And some of them can plot the effective vertical stress distribution as well as the water pressure distribution over the depth.  To let the test scripts generate such plots, run them in a shell where the environment variable `KRATOS_G ...
```
Why are the sentences followed by 2 blanks i.s.o. 1

avdg81 replied:
It is a convention that I have seen somewhere, and which I liked. It helps me to quickly see where one sentence ends and the next one starts. This is particularly useful when reading the documentation in a text editor with a monospace font. We could discuss internally whether we would like to follow this convention, or that we feel it doesn't have any added value.

WPK4FEM commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_processes/test_fix_water_pressures_above_phreatic_line.cpp:
```diff
@@ -193,8 +192,8 @@ KRATOS_TEST_CASE_IN_SUITE(TestFixWaterPressureAbovePhreaticLine_FreesNodesWhenTh
-    // Move node 1 below phreatic line using total displacements (without moving the mesh, since move_mesh is false)
-    r_model_part.GetNode(1).FastGetSolutionStepValue(TOTAL_DISPLACEMENT)[1] = -1.0;
+    // Move node 1 below phreatic line
+    r_model_part.GetNode(1).Coordinates()[1] = -1.0;
```
```suggestion
    r_model_part.GetNode(1).Y() = -1.0;
```
Easier to read for me, because it avoids thinking about indices of directions.

avdg81 replied:
I completely agree. Changed that.

WPK4FEM commented on applications/GeoMechanicsApplication/tests/dsettlement/fully_saturated_column_uniform_load/README.md:
```diff
@@ -12,7 +12,7 @@ The test is performed in 5 stages:
 2. A settlement stage with abc model (100 days or 8640000 seconds). The keywords "reset_displacement" is set to true here, such that the total displacements start counting from the start of this stage.
 3. Another settlement stage (8.64 seconds), the same uniform load and also the same material as previous stage is applied. The time-step is 0.001 days (8.64 seconds), such that there is horizontal stress distribution by the abc model. 
-4. A second settlment stage (1 day) with a uniform line-load of 20 [kN] applied in the negative Y direction to the top of the model.
+4. A second settlement stage (1 day) with a uniform line-load of 20 [kN] applied in the negative Y direction to the top of the model.
```
Elsewhere, kN was changed to kPa. After all it is a distributed load. The same applies in line 16.

avdg81 replied:
Thanks for pointing this out. It has been corrected now.

WPK4FEM commented on applications/GeoMechanicsApplication/tests/dsettlement/phreatic_line_below_soil_surface/MaterialParameters_linear.json:
```diff
@@ -0,0 +1,34 @@
+                    "MINIMUM_RELATIVE_PERMEABILITY": 0.0001,
+                    "K0_MAIN_DIRECTION": 1,
+                    "K0_NC": 0.6,
+                    "YOUNG_MODULUS": 1.0e9,
```
At some point, we may consider a natural order of these parameters. For this linear elastic law, the elastic parameters (Young's modulus and Poisson's ratio) are the definition of the material. I would expect those to be among the first mentioned.
This is a discussion point for the future and may be disregarded here for the time being.

avdg81 replied:
This is a valid point. I do, however, agree with your suggestion to discuss this first. If we then feel that we need to modify it, we may want to apply the same change to other tests, too.

WPK4FEM commented on applications/GeoMechanicsApplication/tests/dsettlement/phreatic_line_below_soil_surface/ProjectParameters_stage1.json:
```diff
@@ -0,0 +1,223 @@
+        },
+        "time_stepping": {
+            "time_step": 1,
+            "max_delta_time_factor": 86400000000
```
The factor given seems to have to do with a conversion between days and seconds. The meaning of max_delta_time_factor is the maximum ratio between an upscaled time step size and the original step size provided by the user. That makes this a strange value. I have informed Walter about these.

avdg81 replied:
I'm not sure what to do this with this comment. If you would like me to make any changes, can you please indicate what those should be, in your opinion? Thank you.

rfaasse commented on applications/GeoMechanicsApplication/custom_processes/fix_water_pressures_above_phreatic_line.cpp:
```diff
@@ -42,9 +42,7 @@ FixWaterPressuresAbovePhreaticLineProcess::FixWaterPressuresAbovePhreaticLinePro
-        const auto current_coordinates =
-            rNode.GetInitialPosition() + rNode.FastGetSolutionStepValue(TOTAL_DISPLACEMENT);
-        if (current_coordinates[1] > mPhreaticLineTable(current_coordinates[0])) {
+        if (rNode.Coordinates()[1] > mPhreaticLineTable(rNode.Coordinates()[0])) {
```
Ahh of course, we can just use the current position, then we don't need the move_mesh flag either (and it's nice that total displacement is now once more only an output item).

rfaasse commented on applications/GeoMechanicsApplication/tests/moving_column_with_fixed_pressure_above_phreatic_line.py:
```diff
@@ -43,7 +43,7 @@ def test_fixed_pressure_above_phreatic_line_via_cpp_workflow_without_move_mesh(
         self.assertEqual(status, 0)
 
-        self.assert_results(test_folder, expected_value1 = -4479.17, expected_value2 = -4270.83)
+        self.assert_results(test_folder, expected_value1 = 0.0, expected_value2 = 0.0)
```
To avoid confusion, could you also adjust the comment at the top of the test, as well as in line 69, as well as the README.md for this test, such that it's clear in the future why this one now expects 0.0?

avdg81 replied:
We have updated the comments and documentation together.

WPK4FEM commented on applications/GeoMechanicsApplication/tests/test_helper.py:
```diff
@@ -618,6 +637,24 @@ def element_integration_point_values_at_time(result_item_name, time, output_data
             return element_results
 
 
+def extract_nodal_settlement_over_time(output_data, node_id):
```
settlement in the function name suggests that this function belongs to a specific test and should be placed in the python script for that specific test.
In the function body, the multiplication with -1 and the seconds_to_days conversion also suggest that this is specific rather than general functionality.

avdg81 replied:
I see your point. Let's move this function back to the test script (where it was while I was developing the test case). We can always see later on whether we want to move it elsewhere if there is a need for it.

rfaasse commented on applications/GeoMechanicsApplication/python_scripts/run_geo_settlement.py:
```diff
@@ -20,18 +22,19 @@ def run_stages(working_dir, project_parameters_filenames) -> int:
-        if status != 0:
-            print(f"Running analysis stage failed: status = {status}")
-            return status
+    with context_managers.set_cwd_to(working_dir):
```
Nice, we can start using this in multiple places (I didn't know we had this).

WPK4FEM commented on applications/GeoMechanicsApplication/tests/test_dsettlement_validation.py:
```diff
@@ -1,9 +1,140 @@
+        [
+            stress_vector[1]
+            for stress_vector in test_helper.GiDOutputFileReader.nodal_values_at_time(
+                "CAUCHY_STRESS_TENSOR", time_in_s, output_data, node_ids=node_ids
```
Probably my lack of understanding:
Why give node_ids=node_ids as last argument to GiDOutputFileReader.nodal_values_at_time?
I would expect just node_ids

avdg81 replied:
`node_ids` is a [keyword argument](https://docs.python.org/3/tutorial/controlflow.html#keyword-arguments), which you would normally use in the way shown here. A nice property of keyword arguments is that their positions in the argument list don't matter, since they are supplied through their names contrary to positional arguments. So the first occurrence of `node_ids` refers to the name of the keyword argument, whereas the second occurrence refers to the actual value for this argument (which happens to have the same name).

Does this adequately answer your question?

WPK4FEM commented on applications/GeoMechanicsApplication/tests/test_dsettlement_validation.py:
```diff
@@ -1,9 +1,140 @@
+def get_nodal_water_pressures_at_time(time_in_s, output_data, node_ids=None):
+    return make_compression_positive_and_convert_Pa_to_kPa(
+        test_helper.GiDOutputFileReader.nodal_values_at_time(
+            "WATER_PRESSURE", time_in_s, output_data, node_ids=node_ids
```
same question as on line 37.

avdg81 replied:
See my previous reply. :-)

WPK4FEM commented on applications/GeoMechanicsApplication/tests/test_dsettlement_validation.py:
```diff
@@ -1,9 +1,140 @@
+
+    # Extract data points from the Kratos analysis results
+    water_pressures = get_nodal_water_pressures_at_time(
+        time_in_s, output_data, node_ids=node_ids_over_depth
```
Same question:
Isn't node_ids_over_depth sufficient as last argument to pass?

avdg81 replied:
See my previous reply. :-)

WPK4FEM commented on applications/GeoMechanicsApplication/tests/test_dsettlement_validation.py:
```diff
@@ -1,9 +1,140 @@
+    data_series_collection.append(
+        plot_utils.DataSeries(
+            zip(effective_vertical_stresses, y_coordinates, strict=True),
+            "sigma_yy;eff [Kratos]",
```
;eff indicates that the stress is an effective stress. The label now consists of the greek word i.s.o. a greek symbol, followed  by _yy to indicate the component and ;eff for effective. Using ; seems unusual.

Maybe the label becomes long, but the following would be clearer for me:
"effective stress_yy [Kratos]"

avdg81 replied:
I have found that we can use LaTeX syntax for the labels, which allows me to write Greek letters and subscripts. Let me know whether you like that more. Thanks.

WPK4FEM replied:
The use of Greek symbols nicely improves readability. Thank you.

WPK4FEM commented on applications/GeoMechanicsApplication/tests/test_dsettlement_validation.py:
```diff
@@ -31,48 +159,57 @@ def test_settlement_dry_column(self):
-            abs((
-                expected_settlement_after_one_hundred_days
-                - actual_settlement_after_one_hundred_days
+            abs(
+                (
+                    expected_settlement_after_one_hundred_days
+                    - actual_settlement_after_one_hundred_days
+                )
+                / expected_settlement_after_one_hundred_days
```
Quite lengthy formatting, is this what black gives?

avdg81 replied:
Yes, this is indeed how `black` formats this piece of code.

By the way, I've noticed that this calculation of the relative error is repeated several times. We may want to extract that to a separate function to simplify the code and reduce code duplication. I'd suggest to do that in a separate PR, to avoid adding even more changes to this PR.

WPK4FEM replied:
o.k.

WPK4FEM commented on applications/GeoMechanicsApplication/tests/test_dsettlement_validation.py:
```diff
@@ -31,48 +159,57 @@ def test_settlement_dry_column(self):
-            abs((
-                expected_settlement_after_ten_thousand_days
-                - actual_settlement_after_ten_thousand_days
+            abs(
```
The computation of a relative difference repeats here. COuld be a candidate to place in test_helper as a general function.

avdg81 replied:
Yes, I had the same idea (see my previous reply).

WPK4FEM commented on applications/GeoMechanicsApplication/tests/test_dsettlement_validation.py:
```diff
@@ -104,22 +238,29 @@ def test_settlement_consolidation_coarse_mesh(self):
         nodes = [3, 15, 16, 17, 18, 4]
         expected_pressure = [0.0, -18524.3, -24096.0, -33456.2, -42816.6, -50000.0]
-        expected_total_stress = [-5332.28, -24858.1, -52279.1, -70030.9, -89823.8, -109627.0]
-        
+        expected_total_stress = [
+            -5332.28,
+            -24858.1,
+            -52279.1,
+            -70030.9,
+            -89823.8,
+            -109627.0,
+        ]
+
```
formatting of expected_pressure is as a row, expected_total_stress as a column.
Why? Is it the trailing , on the last entry that causes a column?
The same pattern repeats after line 292

avdg81 replied:
This is a result of running `black`. I think it's because the numbers contain many digits, so having them all on a single line would exceed the maximum line width applied by `black`. Shall we simply accept this, or do you have a suggestion on how we can make this better?

WPK4FEM replied:
Sometimes we block formatting for a section like this. However that also draws attention to itself. Here we can leave the formatting to black. What still draws my attention is the inconsistency in the trailing comma.

rfaasse commented on applications/GeoMechanicsApplication/tests/dsettlement/phreatic_line_below_soil_surface/test_case_3_settlement_plot.svg:
Looks good!

WPK4FEM commented on applications/GeoMechanicsApplication/tests/test_dsettlement_validation.py:
```diff
@@ -131,9 +272,6 @@ def test_settlement_consolidation_fine_mesh(self):
-        current_working_directory = os.getcwd()
-        os.chdir(project_path)
-
         import KratosMultiphysics.GeoMechanicsApplication.run_geo_settlement as run_geo_settlement
```
Apart from the expected result numerical values, the fine mesh and the coarse mesh test seem to be modified copies of eachother

avdg81 replied:
I assume you are suggesting to remove this duplication in a separate PR...?

WPK4FEM replied:
yes

WPK4FEM reviewed (commented):
Dear Anne,
Thank you, this has become quite an extensive pull request. The picture generation is very useful for judgement of the results. We should consider to add pictures to the READ.me files of the other than no 3 test cases in a next pull request. Please find my other remarks here.
Regards, Wijtze Pieter

rfaasse commented on applications/GeoMechanicsApplication/tests/test_helper.py:
```diff
@@ -438,6 +440,23 @@ def are_dictionaries_almost_equal(expected: Dict[Any, Any],
+    return os.environ.get("KRATOS_GEO_MAKE_TEST_PLOTS", "off").lower() == "on"
+
+
+def get_data_points_from_file(file_path, line_to_data_point):
```
I had to read this function a few times before I understood that the line_to_data_point is a function. We could rename the variable or e.g. give a type hint here to help future readers

avdg81 replied:
That is a fair point. I have renamed the variable to `data_point_extractor`. Do you feel that is a better name?

rfaasse commented on applications/GeoMechanicsApplication/tests/test_dsettlement_validation.py:
```diff
@@ -190,25 +333,274 @@ def test_settlement_fully_saturated_column(self):
         self.assertAlmostEqual(actual_settlement_after_ten_thousand_days, -8.63753, 4)
 
-        os.chdir(cwd)
+        if test_helper.want_test_plots():
```
The code within this if-statement is almost identical to the blocks for the other two tests. Is it possible to combine them?

avdg81 replied:
The problem that I see here is that there are several subtle differences in these pieces of test code. For instance, the times at which we generate stress plots slightly differs from test case to test case. Furthermore, it depends on the test case which stages need to be considered for the plot. For one test case it is the last three stages, for another one it is the last four stages. I haven't found a clear way to capture these differences. If you have an idea, please let me know. Otherwise, I'd suggest to leave it the way it is now. Thanks.

rfaasse replied:
Hmm okay, then indeed I think it's okay to leave it as is

rfaasse commented on applications/GeoMechanicsApplication/tests/test_dsettlement_validation.py:
```diff
@@ -190,25 +333,274 @@ def test_settlement_fully_saturated_column(self):
         )
         self.assertEqual(status, 0)
 
+        project_path = pathlib.Path(project_path)
```
Since we now need this, couldn't we make sure the `get_file_path` function in line 323 returns a Path instead of a string?

avdg81 replied:
That is an excellent suggestion. However, I feel this is a step too far for this PR, since we have to make sure that no client code assumes that a string is returned rather than a path-like object. Let's create a new issue to further work out this idea.

rfaasse commented on applications/GeoMechanicsApplication/python_scripts/geo_plot_utilities.py:
```diff
@@ -0,0 +1,60 @@
+
+    for series in data_series_collection:
+        axes.plot(
+            *zip(*series.data_points),
```
This is a lack in my understanding, but this line is not completely clear to me: what is the type of data_points (so related, what is the output of the zip)? And why do we need the two `*`.

avdg81 replied:
Here `data_points` is assumed to be an iterable that contains pairs of numbers, where each pair represents a point that needs to be plotted. I've taken this code from the following discussion on [StackOverflow](https://stackoverflow.com/questions/21519203/plotting-a-list-of-x-y-coordinates).

rfaasse commented on applications/GeoMechanicsApplication/python_scripts/geo_plot_utilities.py:
```diff
@@ -0,0 +1,60 @@
+        axes.set_ylabel(ylabel)
+    figure.legend(loc="outside center right")
+
+    if isinstance(plot_file_path, pathlib.Path):
+        plot_file_path = str(plot_file_path.resolve())
+    print(f"Saving plot to {plot_file_path}")
+    plt.savefig(plot_file_path)
```
Wat will happen if the path doesn't exist here?

avdg81 replied:
I'm not entirely sure. When the parent folder of the image file exists, it will automatically create (or overwrite) this file. If, however, some subdirectories in the given `plot_file_path` don't exist, I'm not sure whether they will be created as well. It's very well possible this is not the case. I'd suggest to address this potential issue when the need arises, since so far we put the plot files in the test directory (which must exist, or else we couldn't run the test in the first place).

rfaasse commented on applications/GeoMechanicsApplication/python_scripts/geo_plot_utilities.py:
```diff
@@ -0,0 +1,60 @@
+    )
+
+
+def make_stress_over_y_plot(data_series_collection, plot_file_path):
```
Is this a common name for stress plots? For me personally, I think something like `make_stress_depth_plot` is more intuitive

avdg81 replied:
I'm not sure what is the best name for this function. Perhaps @WPK4FEM has a suggestion?

rfaasse commented on applications/GeoMechanicsApplication/tests/dsettlement/phreatic_line_below_soil_surface/MaterialParameters_linear.json:
```diff
@@ -0,0 +1,34 @@
+                    "name": "GeoLinearElasticPlaneStrain2DLaw"
+                },
+                "Variables": {
+                    "IGNORE_UNDRAINED": true,
```
Since we're dealing with water in the other material parameters, should this one also be set to `false`?

rfaasse replied:
Ah I see that the water is fixed completely using a hydrostatic profile in stage 1 (if I understand correctly), keeping this comment to validate my understanding

avdg81 replied:
Yes, your second comment is correct.

rfaasse commented on applications/GeoMechanicsApplication/tests/dsettlement/phreatic_line_below_soil_surface/README.md:
```diff
@@ -0,0 +1,54 @@
+
+![Settlement](test_case_3_settlement_plot.svg)
+
+Stresses and water pressures after 100 days of settlement under self weight:
```
To avoid ambiguity, could you add here where in the column the data for the depth plot is coming from (i.e. left, right, middle)?

avdg81 replied:
Added.

rfaasse commented on applications/GeoMechanicsApplication/tests/dsettlement/phreatic_line_below_soil_surface/test_case_3_stress_plot_after_100.1_days.svg:
Nitpicking, for consistency, you could consider making the two groups of legends more similar, so ref P_w vs Kratos P_w or P_w [ref] vs P_w[Kratos]. To me also the x label is not completely consistent with the water pressure graph (although of course the units are the same), but I'm not sure what would be better

avdg81 replied:
Based on the feedback that I received from @WPK4FEM, the labels have been improved. Let me know if you see more room for improvement.

rfaasse commented on applications/GeoMechanicsApplication/tests/dsettlement/phreatic_line_below_soil_surface/README.md:
```diff
@@ -0,0 +1,54 @@
+## Assertions
+
+The following assertions are made in this test (assuming the results from D-Settlement):
+1. The total vertical displacement at the top of the column is expected to be $`0 [\mathrm{m}]`$ after about $`0.1 [\mathrm{d}]`$ (stage 3).
```
Is this a typo? 
```suggestion
1. The total vertical displacement at the top of the column is expected to be $`0 [\mathrm{m}]`$ after about $`0.1 [\mathrm{d}]`$ (stage 2).
```

avdg81 replied:
No, this is not a typo. At the start of stage 3, we reset the total displacement field. Therefore, we expect the settlement to be nearly equal to 0 in the first time step of that stage. Note that how the stages have been defined for the various settlement test cases slightly differs per case.

rfaasse commented on applications/GeoMechanicsApplication/tests/test_dsettlement_validation.py:
```diff
@@ -1,9 +1,140 @@
+    import KratosMultiphysics.GeoMechanicsApplication.geo_plot_utilities as plot_utils
+
+
+def _extract_x_and_y_from_line(line, index_of_x=0, index_of_y=1):
```
Since this function isn't in a class, should the _ be removed at the start?
```suggestion
def extract_x_and_y_from_line(line, index_of_x=0, index_of_y=1):
```

avdg81 replied:
I'm in the habit of prepending "private" function names with an underscore. But perhaps we should have a discussion with the team about this, to make sure we all use the same convention. Thanks for bringing this up.

rfaasse replied:
Yes, as discussed, I wasn't used to this concept of 'private free functions', but it makes sense (especially since we don't have things like anonymous namespaces to make these functions private if we'd like to in python.

avdg81 committed:
Processed Wijtze Pieter's review suggestions

- For readability, use member function `Y()` rather than `Coordinates()[1]`.
- Corrected the unit of a surface load (kN -> kPa).
- Use greek symbols and subscripts for some of the labels.
- Moved a function from one Python file to another one.
- Fixed plotting the grid lines.

mnabideltares commented on applications/GeoMechanicsApplication/tests/dsettlement/phreatic_line_below_soil_surface/README.md:
```diff
@@ -0,0 +1,54 @@
+# Settlement of a soil column with the phreatic line below the soil surface
+
+This test is based on Test Case 3.  The model consists of a rectangular soil domain, $`1 [\mathrm{m}]`$ wide and $`50 [\mathrm{m}]`$ deep.  The mesh consists of `SmallStrainUPwDiffOrderElement2D6N` elements.  The phreatic line is located $`10 [\mathrm{m}]`$ below the soil surface, namely at $y = 40 [\mathrm{m}]$. Assuming a hydrostatic pressure distribution, the pressure boundary at the bottom is ...
```
40 m deep. Doesn't the pressure become 400 kPa, rather than 40?

avdg81 replied:
Good catch! Yes, you're completely right. Corrected that.

mnabideltares commented on applications/GeoMechanicsApplication/tests/dsettlement/phreatic_line_below_soil_surface/README.md:
```diff
@@ -0,0 +1,54 @@
+- Displacements on the bottom are fixed in all directions.
+- Displacements on the sides are fixed in the $`X`$ direction.
+- Gravity is applied to the entire domain ($`-9.81 [\mathrm{m}/\mathrm{s}^2]`$ in the negative $`Y`$ direction).
+- Water pressures at the bottom are fixed at $`40 [\mathrm{kPa}]`$.
```
40 or 400? Please see the comment above

avdg81 replied:
Corrected it here, too. Thanks for pointing this out.

rfaasse commented on applications/GeoMechanicsApplication/tests/test_dsettlement_validation.py:
```diff
@@ -31,48 +159,57 @@ def test_settlement_dry_column(self):
         )
         self.assertEqual(status, 0)
 
+        project_path = pathlib.Path(project_path)
```
I don't know if it would be too much work, but couldn't the test_helper in line 149 return a Path instead of a string?

avdg81 replied:
See one of my earlier replies.

rfaasse commented on applications/GeoMechanicsApplication/tests/test_dsettlement_validation.py:
```diff
@@ -190,25 +333,274 @@ def test_settlement_fully_saturated_column(self):
+                "time_in_s": unit_conversions.days_to_seconds(0.1) + 1.0,
+                "expected_total_u_y": 0.0,
+                "delta": 0.01,
+            },
```
Nice, this is easy to read

rfaasse commented on applications/GeoMechanicsApplication/tests/test_dsettlement_validation.py:
```diff
@@ -190,25 +333,274 @@ def test_settlement_fully_saturated_column(self):
+            {
+                "output_data": output_stage_3,
+                "time_in_s": unit_conversions.days_to_seconds(100),
+                "expected_total_u_y": -1.75,
```
You could add here that this is an analytical value (since you added a comment about regression values for the others)

avdg81 replied:
Done.

rfaasse reviewed (commented):
Thanks for adding this test and making improvements to settlement functionality! I'd suggest to have the title of the PR reflect a bit more than only adding test case 3 (since a lot more happened here). You'll find my other suggestions/questions in the comments, but none of them are major or blockers blockers

mnabideltares commented on applications/GeoMechanicsApplication/python_scripts/unit_conversions.py:
```diff
@@ -0,0 +1,10 @@
+def seconds_to_days(duration_in_seconds):
+    return duration_in_seconds / (24.0 * 60.0 * 60.0)
```
It is matter of taste, but I think dividing by 86400 is more clear, and it avoids extra calculations (although very tiny)

avdg81 replied:
I have now defined a variable `number_of_seconds_per_day` and set its value to 86400. This new variable is now used by the conversion functions you've mentioned.

mnabideltares commented on applications/GeoMechanicsApplication/python_scripts/unit_conversions.py:
```diff
@@ -0,0 +1,10 @@
+
+
+def days_to_seconds(duration_in_days):
+    return duration_in_days * 24.0 * 60.0 * 60.0
```
Same as comment above

avdg81 replied:
See my earlier reply.

mnabideltares commented on applications/GeoMechanicsApplication/tests/dsettlement/phreatic_line_below_soil_surface/ProjectParameters_stage3.json:
```diff
@@ -0,0 +1,204 @@
+                            "body_output": true,
+                            "node_output": false,
+                            "skin_output": false,
+                            "nodal_results": ["TOTAL_DISPLACEMENT", "WATER_PRESSURE", "CAUCHY_STRESS_TENSOR"],
```
CAUCHY_STRESS_TENSOR is given as output on nodes in stages 3, 4 and 5. But it is not the case in stage 2. Is there any reason for this?

avdg81 replied:
The stages that ask for output item `CAUCHY_STRESS_TENSOR` are relevant for the stress plots, since they contain time steps that are important for the comparison with D-Settlement results. That is not the case for stage 2. Therefore, we left it out from that stage.

mnabideltares reviewed (commented):
@avdg81 
Thank you for adding case 3, and making graphs for all cases. It helped us to validate the cases in more details.
I find, it is implemented nicely and clean. I only have minor comments.

avdg81 committed:
Updated the generated plots

Some plots now use Greek letters and subscripts in their labels.

avdg81 committed:
Merge remote-tracking branch 'origin/master' into geo/13745-settlement-test-case-3

avdg81 committed:
Processed Richard's review suggestions

- The description of the test where the soil column moves up and down has been modified such that it reflects the changes made to the process that fixes the water pressures above the phreatic line.

avdg81 committed:
Continued to process Richard's review suggestions

- Renamed a function argument to clarify its meaning.
- Clarified along which cross-section the stress plots have been taken.
- Annotated a few expected values by saying that they have been taken from analytical results.

avdg81 committed:
Merge remote-tracking branch 'origin/master' into geo/13745-settlement-test-case-3

avdg81 committed:
Reverted commenting out a failing assertion

Due to Aron's fix, this assertion no longer fails.

avdg81 committed:
Processed Mohamed's review suggestions

- Corrected the documented water pressure at the bottom of the model.
- Defined a variable that stores the number of seconds per day, and used it in the duration conversion functions.

rfaasse reviewed (dismissed):
Thank you for processing the review comments! To me this looks ready to go. To me it makes sense if Mohamed or Wijtze Pieter also gives the green light from the computational side, but I leave that up to you

avdg81 committed:
Forgot to add a comment

The added comment explains that data is unpacked from pairs into two lists.

rfaasse reviewed (approved):
Thanks for adding clarification in a comment, still ready to go!
